### ISSUE TITLE
No binding for g_dir_open

### ISSUE THREAD

**Issue by psychon** [pavouk/lgi#135]
Hi,

I'm trying to access `g_dir_open`, but I haven't found it anywhere accessible. Thus, I came up with the following.
Some questions:

- Is the code for `g_dir_open` correct?
- Why does the code for using `g_dir_close` as `__gc` not work? How can it be made to work?
- Is this correct enough for being added to `lgi/override/GLib.lua`?

```lua
local core = require("lgi.core")
local ti = require("lgi.ffi").types
local dir = require("lgi").GLib.Dir
local glib_module = core.module("glib-2.0")

-- Add GLib.dir.open() and use it as the constructor for dirs
dir._method.open = core.callable.new {
	name = "GLib.Dir.open",
	addr = assert(glib_module.g_dir_open),
	ret = dir,
	-- Arguments
	ti.filename,
	ti.int,
	-- TODO: Third argument is GError**, how? Is the following correct?
	throws = true
}
dir._new = function(_, path) return dir._method.open(path, 0) end

-- We don't want no memory leaks
-- TODO: According to valgrind, this does not work
dir._free = dir.close

-- Test if this works
local d = dir("/tmp")
repeat
	local f = d:read_name()
	print(f or "<done>")
until f == nil

print("Does not exist:")
print(dir("does not exist"))
```
Thanks,
Uli

Edit: I just wondered: Would it make sense to have something like the following available as lgi.GLib.something?
```lua
function iterate_over_dir(path)
	local dir, err = dir(path)
	if err then error(err) end
	return dir.read_name, dir
end
for file in iterate_over_dir("/tmp") do
	print(file)
end
```

**Comment by pavouk** [comment]
Hi,

> On 27 Apr 2016, at 11:28, Uli Schlachter <notifications@github.com> wrote:
> 
> Hi,
> 
> I'm trying to access g_dir_open, but I haven't found it anywhere accessible. Thus, I came up with the following.
> Some questions:
> 
> Is the code for g_dir_open correct?
Yes, perfect.
> Why does the code for using g_dir_close as on garbage collection not work? How can it be made to work?
_free expects either lightuserdata with raw pointer to C function, or gi.c’s userdata with method definition.  I guess that in your case, you can use something like core.gi.GLib.Dir.methods.close (see GLib-Variant.lua override for reference).
> Is this correct enough for being added to lgi/override/GLib.lua?
Yes it is correct.  However, I’m a bit reluctant to add more hacks for ancient GLib facilities functionally replaced by binding-friendly Gio.  Is there any problem with Gio.File and Gio.FileEnumerator that make it unusable for your task?

Pavel

**Comment by psychon** [comment]
The only problem I have with Gio.File and Gio.FileEnumerator is that I didn't find them. For some reason I only looked into glib. Thanks for the pointer!

Closing this since a better solution exists.

**Comment by Elv13** [comment]
>  Is there any problem with Gio.File and Gio.FileEnumerator that make it unusable for your task?

Well, https://github.com/pavouk/lgi/issues/88, as far as I know (I did not test with 0.9 yet) is still a big issue (to me and many other Awesome users). It happens when there is multiple layers of recursive async calls, then LGI has a significant chance of crashing with an use after free. Last time I tried, it could still be reproduced using any of the "demo" from my https://github.com/Elv13/wirefu library (a wrapper to make dbus easier to use). The issue has also been reported by people who forked [this](https://github.com/Elv13/awesome-configs/blob/master/utils/fd_async.lua) to use it in their LGI project. I use it, but it is only stable as long as there is no recursive calls. My workaround is to queue them using GMainLoop run_idle. I know it is an heisenbug and is hard to reproduce and debug, it will happen, eventually (unless you fixed it and I failed to notice).
